File: two_pointers/2799.py
# Tc = O(n) [each element visiting twice only]
# Sc = O(n)

from collections import defaultdict

class Solution:
    def countCompleteSubarrays(self, nums):
        n = len(nums)
        unique_elements = set(nums)
        c = len(unique_elements)
        i = 0
        j = 0
        res = 0
        mp = defaultdict(int)
        while j < n:
            mp[nums[j]] += 1
            while len(mp) == c:
                res += (n - j)
                mp[nums[i]] -= 1
                if mp[nums[i]] == 0:
                    del mp[nums[i]]
                i += 1
            j += 1
        return res

File: two_pointers/2799.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countCompleteSubarrays = function (nums) {
    const n = nums.length;
    const uniqueElements = new Set(nums);
    const c = uniqueElements.size;
    let i = 0;
    let j = 0;
    let result = 0;
    const mp = new Map();

    while (j < n) {
        mp.set(nums[j], (mp.get(nums[j]) || 0) + 1)
        while(mp.size === c){
            result += (n - j)
            mp.set(nums[i], mp.get(nums[i]) - 1)
            if(mp.get(nums[i]) === 0){
                mp.delete(nums[i])
            }
            i += 1
        }
        j += 1
    }
    return result
};